### PR TITLE
Add proxy agent runner for kubeflow pipeline

### PR DESCRIPTION
**Pull request by IronPan**
This is the initial change to add a [proxy agent](https://github.com/google/inverting-proxy) for a kubeflow pipeline cluster. 

The container should be launched as a K8s Deployment in the same namespace that kubeflow is deployed. 

The agent runner will register to the proxy server and store the metadata in a ConfigMap. In a failover case that the agent die, the new agent should pick up the metadata from the ConfigMap.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/988)

### CODE REVIEW

File: proxy/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Ping to a specific version of invert proxy agent
```
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/Dockerfile]: True. I start with Pipeline just to make iteration and release fast (weekly cadence). Once stable, we are free to move it to main repo. (Moving the code is low cost)

[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/Dockerfile]: yeah that's the plan. here is the PR https://github.com/kubeflow/pipelines/pull/996

[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/Dockerfile]: done

File: proxy/attempt-register-vm-on-proxy.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        --shim-path=websocket-shim \
+        --health-check-path=${HEALTH_CHECK_PATH} \
+        --health-check-interval-seconds=${HEALTH_CHECK_INTERVAL_SECONDS} \
+        --health-check-unhealthy-threshold=${HEALTH_CHECK_UNHEALTHY_THRESHOLD}
+}
+
+# Check if the cluster already have proxy agent installed by checking ConfigMap.
+# If so, reuse the existing endpoint (a.k.a BACKEND_ID).
+if kubectl get configmap inverse-proxy-config; then
+  PROXY_URL=$(kubectl get configmap inverse-proxy-config -o json | jq -r ".data.ProxyUrl")
```
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/attempt-register-vm-on-proxy.sh]: done

File: proxy/get_proxy_url.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  urls = urls_for_zone(location, regional_urls)
+  if not urls:
+    raise ValueError("No valid URLs found for zone: {}".format(location))
+
+  for url in urls:
+    try:
+      status_code = requests.head(url).status_code
+    except requests.ConnectionError:
+      pass
+    expected_codes = frozenset([307, 401])
```
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/get_proxy_url.py]: thanks! 307 is legit. 401 is not good. I'll remove it.

File: proxy/get_proxy_url.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    # 307 - Temporary Redirect, Jupyter sends this if VM has access rights.
+    # 401 - unauthorized, proxy sends this if VM does not have access rights.
+    if status_code in expected_codes:
+      logging.debug("Status code from the url %s", status_code)
+      print(url)
+      return
+    logging.debug("Incorrect status_code from the server: %s. Expected: %s",
+                  status_code, expected_codes)
+  raise ValueError("No working URL found")
```
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/get_proxy_url.py]: no pod can't find a url so it should just terminate.

File: proxy/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/Dockerfile]: done

File: proxy/attempt-register-vm-on-proxy.sh [hunk 5]
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash
+#
+# Copyright 2017 Google Inc.
```
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/attempt-register-vm-on-proxy.sh]: done

File: proxy/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (hongye-sun) [reply proxy/Dockerfile]: sg

File: proxy/get_proxy_url.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hongye-sun) [reply proxy/get_proxy_url.py]: sg

File: proxy/get_proxy_url.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  if not urls:
+    raise ValueError("No valid URLs found for zone: {}".format(location))
+
+  for url in urls:
+    try:
+      status_code = requests.head(url).status_code
+    except requests.ConnectionError:
+      pass
+    expected_codes = frozenset([307])
+    # 307 - Temporary Redirect, Proxy server sends this if VM has access rights.
```
Comment (hongye-sun) [inline proxy/get_proxy_url.py]: nit: it's still weird that the endpoint only returns 307 as normal response status and not 200. Could you confirm if it's the case with the Jaas team?

Comment (IronPan) [reply proxy/get_proxy_url.py]: yeah i double checked with slava earlier and this is intended behavior.

**Comment by IronPan** [comment]
wait for @ojarjur review

File: proxy/get_proxy_url.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    if status_code in expected_codes:
+      logging.debug("Status code from the url %s", status_code)
+      print(url)
+      return
+    logging.debug("Incorrect status_code from the server: %s. Expected: %s",
+                  status_code, expected_codes)
+  raise ValueError("No working URL found")
+
+
+main()
```
Comment (Ark-kun) [inline proxy/get_proxy_url.py]: Usually the `main()` call is put inside
```
if __name__ == '__main__':
```
This way the module can be loaded and used like a library.

File: proxy/get_proxy_url.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  urls = urls_for_zone(location, regional_urls)
+  if not urls:
+    raise ValueError("No valid URLs found for zone: {}".format(location))
+
+  for url in urls:
+    try:
+      status_code = requests.head(url).status_code
+    except requests.ConnectionError:
+      pass
+    expected_codes = frozenset([307])
```
Comment (Ark-kun) [inline proxy/get_proxy_url.py]: JFYI, the plain `set` also exists.
Nothing wrong with `frozenset` (immutable) though.

File: proxy/get_proxy_url.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""CLI tool that returns URL of the proxy for particular zone and version."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
Comment (Ark-kun) [inline proxy/get_proxy_url.py]: JFYI, These imports are probably only needed for python 2.

File: proxy/get_proxy_url.py (hunk 6 above)
Comment (Ark-kun) [reply proxy/get_proxy_url.py]: 3xx codes are pretty normal.
Codes < 400 describe success.

**Comment by Ark-kun** [comment]
Some general notes:
* Is the root directory the best place for this program? Maybe we can have something like `deployment` directory and put all deployment-related files there.
* From user's perspective, how are they supposed to use/launch this?

File: proxy/get_proxy_url.py (hunk 7 above)
Comment (IronPan) [reply proxy/get_proxy_url.py]: done

File: proxy/get_proxy_url.py (hunk 8 above)
Comment (IronPan) [reply proxy/get_proxy_url.py]: ack. thanks

File: proxy/get_proxy_url.py (hunk 9 above)
Comment (IronPan) [reply proxy/get_proxy_url.py]: indeed. removed

File: proxy/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (IronPan) [reply proxy/Dockerfile]: done

**Comment by IronPan** [comment]
added the tests as @ojarjur suggested.

> Is the root directory the best place for this program? Maybe we can have something like deployment directory and put all deployment-related files there.
I think so. Pipeline won't have its own deployment, and i'm considering moving this code to kubeflow repo once stable. Hosting here just for faster release. 
From user's perspective, how are they supposed to use/launch this?

**Pull request merged**